src/mon/MgrMap.h

```diff
@@ -15,30 +15,140 @@
 #define MGR_MAP_H_
 
 #include <sstream>
+#include <set>
 
 #include "msg/msg_types.h"
 #include "common/Formatter.h"
 #include "include/encoding.h"
+#include "include/utime.h"
 #include "common/version.h"
+#include "common/options.h"
+#include "common/Clock.h"
 
 
 class MgrMap
 {
 public:
+  struct ModuleOption {
+    std::string name;
+    uint8_t type = Option::TYPE_STR;         // Option::type_t TYPE_*
+    uint8_t level = Option::LEVEL_ADVANCED;  // Option::level_t LEVEL_*
+    uint32_t flags = 0; // Option::flag_t FLAG_*
+    std::string default_value;
```

Comment:
It's more awkward to use the variants at this layer because the encode/decode infrastructure over the write doesn't understand them.  and at the config level in core ceph, all options are stored as strings.  (they are declared in options.cc as their type for type safety, but ultimately they are sent over the write to daemons etc as strings and converted to their proper type later.)  So i'm inclined to stick with strings here at least..

Reply:
Ops, I assumed they were ultimately packed as compact types.